### PR TITLE
feat: introduce synchronous replicas constraints

### LINKED ISSUE #234: Restrict synchronous standby names to Pods in different AZs
We need allow the user to restrict the synchronous standby names to the Pods which are in a different AZ than the one of the primary node.

This is important when the number of instances is greater than the number of AZs and you just select one synchronous replica. What would happen in that case when the selected synchronous standby will be in the same AZ of the master, and that AZ becomes unavailable?

issue originally made by @leonardoce

### PR DESCRIPTION
**Pull request by armru**
This patch introduces the possibility to restrict the potential pods that can be elected to synchronous replicas by specifying anti-affinity labels 

Closes #234 

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

### CODE REVIEW

**Comment by armru** [comment]
checking if the  design concept passes the e2e tests

File: config/crd/bases/postgresql.cnpg.io_clusters.yaml [hunk 1]
```diff
@@ -2403,6 +2420,24 @@ spec:
                 description: The timestamp when the last request for a new primary
                   has occurred
                 type: string
+              topology:
+                description: Instances topology.
+                properties:
+                  instances:
+                    additionalProperties:
+                      additionalProperties:
```
Comment (jsilvela) [inline config/crd/bases/postgresql.cnpg.io_clusters.yaml:2428]: this looks fishy

File: api/v1/cluster_types.go [hunk 2]
```diff
@@ -283,6 +283,33 @@ const (
 	PhaseApplyingConfiguration = "Applying configuration"
 )
 
+// PodTopologyLabels represent the topology of a Pod. map[labelName]labelValue
+type PodTopologyLabels map[string]string
+
+// hasSameTopology checks if the two topologies have
+// the same labels
+func (topologyLabels PodTopologyLabels) hasSameLabels(instanceTopology PodTopologyLabels) bool {
```
[the remark this answers is not in the document]
Comment (jsilvela) [reply api/v1/cluster_types.go:291]: after more reading, changed to `matchesTopology`

File: config/crd/bases/postgresql.cnpg.io_clusters.yaml (hunk 1 above)
Comment (jsilvela) [reply config/crd/bases/postgresql.cnpg.io_clusters.yaml:2428]: upon inspection, all good

**Review by jsilvela — APPROVED** [review]
tested manually and code looks good.

**Review by sxd — APPROVED** [review]

**Pull request merged**